Add gatekeeper guidance

gemmaguest opened the pull request:
This PR adds a new developer document containing guidance for gatekeepers. This is an initial draft just based on my own experience/assumptions so please do suggest improvements. It is quite verbose because I have included everything I can think of initially, which may be too heavyweight for most daily gatekeeping, but could provide useful insight for those new to it who don't have such broad experience in Mantid. Let me know what you think.

gemmaguest committed:
Add dev doc for gatekeeper advice

gemmaguest commented:
@mantidproject/gatekeepers **Calling all gatekeepers!!** I have attempted to put together some guidance for how to gatekeep. Please let me know what you think. So far this is just a brain dump of my own experience so it would be good to get insight from as many of you as possible.

rosswhitfield commented on dev-docs/source/Gatekeeping.rst:
```diff
@@ -0,0 +1,156 @@
+
+*Be very careful not to push to* ``main``
+
+* When you are made a gatekeeper on the project you will be given **full, unprotected access to the** ``main`` **branch**. This means you lose the normal protection that stops you being able to push directly to the ``main`` branch - you therefore need to be careful at all times that you do not accidentally push to ``main``, e.g. when you are using Git in your terminal or Git GUI.
```
I don't think this is entirely correct. Project **owners** have unprotected access (although they are still protected against rewritng history with `--force`). I believe gatekeekers can't push to `main` without an approved PR with passing status checks?

gemmaguest replied:
I wasn't sure actually and I couldn't find any documentation to confirm one way or the other

gemmaguest replied:
Ok, after a lot of digging:

Gatekeepers have the [admin role](https://github.com/mantidproject/mantid/settings/access), and [this page](https://docs.github.com/en/enterprise-cloud@latest/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches) says "By default, the restrictions of a branch protection rule don't apply to people with admin permissions to the repository". 

In the branch protection settings:
- `Do not allow bypassing the above settings ` is _not_ enabled so we _do_ allow bypassing protection for admins and others  with "bypass branch protection" permission.
- We have `Restrict who can push to matching branches` enabled and gatekeepers _are_ included, so they can push directly to `main`.
- We have `Allow force pushes` and `Allow deletions` disabled though so no one can force push (phew!). 

We might want to reconsider the roles. It seems unnecessary that gatekeepers have full admin access.

rosswhitfield commented on dev-docs/source/Gatekeeping.rst:
```diff
@@ -0,0 +1,156 @@
+Gatekeeping a Pull Request
+==========================
+
+The final step in our :ref:`development workflow <GitWorkflow>` after a :ref:`pull request has been reviewed <ReviewingAPullRequest>` is gatekeeping. This is second review stage where a senior developer / code owner does a final check of the work before merging it to the ``main`` branch.
```
Are there any additional things we need to pay attention to before merging a PR into the `release-next` branch?

sf1919 replied:
We are very careful these days that no new features get merged into `release-next`. It really should only be bugfixes that have been identified as regressions for that release. Anything else should be merged into main for the following release (although it will be available in the nightly after release when we switch back to nightly being published from the main branch). 

I think being aware of where we are in the release sprint is probably a good thing. We reach a point where we are ready to release but are still building the packages and testing so `release-next` exists but is no longer being auto merged into main. As such any merges into `release-next` after this point would be lost entirely.

rosswhitfield replied:
It probably would be good to have a section in this document explaining that.

gemmaguest replied:
Good point, I'll add one

gemmaguest committed:
Add section about destination branch

this contains more detail on main and release-next branch and some other
things to consider

peterfpeterson commented:
@SilkeSchomann Two gatekeepers on a single PR doesn't happen all that often. When it does, it generally turns into an actual discussion between both of them. Most things get merged or "request changes" pretty quickly.

peterfpeterson commented:
@SilkeSchomann Two gatekeepers on a single PR doesn't happen all that often. When it does, it generally turns into an actual discussion between both of them. Most things get merged or "request changes" pretty quickly.

sf1919 reviewed (approved):
I have read through the guidance and found it thorough and very helpful. It certainly is a great addition to our documentation.

I also built the documentation locally and had no warnings. The links all work and it can be found from the index page of the Developer Docs. On viewing the resulting HTML it was formatted correctly.

Given all of this and the fact that the comments to date appear to have been addressed I am approving this PR. Well written and helpful.
